File: mns-core/src/test/java/cz/fb/manaus/core/dao/BetActionDaoTest.java
package cz.fb.manaus.core.dao;

import com.google.common.collect.Ordering;
import cz.fb.manaus.core.model.BetAction;
import cz.fb.manaus.core.model.BetActionType;
import cz.fb.manaus.core.model.MarketPrices;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.RunnerPrices;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.test.CoreTestFactory;
import org.apache.commons.lang3.time.DateUtils;
import org.hibernate.LazyInitializationException;
import org.junit.Test;

import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.OptionalInt;
import java.util.OptionalLong;
import java.util.Set;

import static com.google.common.collect.ImmutableMap.of;
import static cz.fb.manaus.core.test.CoreTestFactory.newMarket;
import static java.util.Collections.singletonMap;
import static java.util.Comparator.comparing;
import static java.util.Optional.empty;
import static org.apache.commons.lang3.time.DateUtils.addHours;
import static org.hamcrest.CoreMatchers.hasItem;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

public class BetActionDaoTest extends AbstractDaoTest {

    @Test
    public void testBetIds() {
        var market = newMarket();
        marketDao.saveOrUpdate(market);
        createAndSaveBetAction(market, new Date(), singletonMap("k1", "XXX"), BET_ID);
        Set<String> ids = betActionDao.getBetActionIds(market.getId(), OptionalLong.empty(), empty());
        assertThat(ids, hasItem(BET_ID));
    }

    @Test
    public void testUpdateBetId() {
        var market = newMarket();
        marketDao.saveOrUpdate(market);
        createAndSaveBetAction(market, new Date(), Collections.emptyMap(), BET_ID);
        var newId = BET_ID + "_1";
        assertThat(betActionDao.updateBetId(BET_ID, newId), is(1));
        assertThat(betActionDao.updateBetId(BET_ID, newId), is(0));
        assertTrue(betActionDao.getBetAction(newId).isPresent());
        assertFalse(betActionDao.getBetAction(BET_ID).isPresent());
    }

    @Test
    public void testSetBetId() {
        var market = newMarket();
        marketDao.saveOrUpdate(market);
        var action = createAndSaveBetAction(market, new Date(), Collections.emptyMap(), null);
        var actionId = action.getId();
        assertThat(betActionDao.setBetId(actionId, "111"), is(1));
        assertThat(betActionDao.get(actionId).get().getBetId(), is("111"));
    }

    @Test
    public void testBetAction() {
        var market = newMarket();
        marketDao.saveOrUpdate(market);
        createAndSaveBetAction(market, new Date(), singletonMap("k1", "XXX"), BET_ID);
        checkCount(market.getId(), OptionalLong.empty(), empty(), 1);
        checkCount(market.getId(), OptionalLong.of(CoreTestFactory.DRAW), empty(), 1);
        checkCount(market.getId(), OptionalLong.of(CoreTestFactory.DRAW + 1), empty(), 0);
        checkCount(market.getId(), OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY), 1);
        checkCount(market.getId(), OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.BACK), 0);
        checkCount(market.getId(), OptionalLong.empty(), Optional.of(Side.LAY), 1);
        checkCount(market.getId(), OptionalLong.empty(), Optional.of(Side.BACK), 0);
    }

    private void checkCount(String marketId, OptionalLong selId, Optional<Side> side, long expectedCount) {
        assertThat(betActionDao.getBetActions(marketId, selId, side).size(), is((int) expectedCount));
        assertThat(betActionDao.getBetActionIds(marketId, selId, side).size(), is((int) expectedCount));
    }

    @Test
    public void testBetActionProperties() {
        var market = newMarket("33", new Date(), CoreTestFactory.MATCH_ODDS);
        marketDao.saveOrUpdate(market);

        createAndSaveBetAction(market, new Date(), singletonMap("k1", "newer"), BET_ID);
        var action = betActionDao.getBetActions("33", OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY)).get(0);
        betActionDao.getBetActions(OptionalInt.empty());
        assertThat(action.getProperties().size(), is(1));
        assertThat(action.getProperties().get("k1"), is("newer"));
    }


    @Test
    public void testBetActionPropertiesOrder() {
        var market = newMarket("33", new Date(), CoreTestFactory.MATCH_ODDS);
        marketDao.saveOrUpdate(market);
        createAndSaveBetAction(market, new Date(), singletonMap("k1", "newer"), BET_ID);
        createAndSaveBetAction(market, addHours(new Date(), -1), singletonMap("k1", "older"), BET_ID + 1);
        var action = betActionDao.getBetActions("33", OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY)).get(0);
        assertThat(action.getProperties().size(), is(1));
        assertThat(action.getProperties().get("k1"), is("older"));

        action = betActionDao.getBetActions("33", OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY)).get(1);
        assertThat(action.getProperties().size(), is(1));
        assertThat(action.getProperties().get("k1"), is("newer"));
        // delete with properties
        marketDao.delete(market.getId());
    }

    @Test
    public void testBetActionPropertiesDelete() {
        var market = newMarket("33", new Date(), CoreTestFactory.MATCH_ODDS);
        marketDao.saveOrUpdate(market);
        createAndSaveBetAction(market, addHours(new Date(), -1), singletonMap("k1", "older"), BET_ID);

        assertTrue(betActionDao.getBetAction(BET_ID).isPresent());
        marketDao.delete(market.getId());
        assertFalse(betActionDao.getBetAction(BET_ID).isPresent());
    }

    @Test
    public void testBetActionPropertiesDuplicity() {
        var market = newMarket("33", new Date(), CoreTestFactory.MATCH_ODDS);
        marketDao.saveOrUpdate(market);
        createAndSaveBetAction(market, new Date(), of("k1", "v1", "k2", "v2"), BET_ID);
        assertThat(betActionDao.getBetActions(OptionalInt.empty()).size(), is(1));
        assertThat(betActionDao.getBetActions("33", OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY)).size(), is(1));
    }

    @Test
    public void testMaxResults() {
        var market = newMarket("33", new Date(), CoreTestFactory.MATCH_ODDS);
        marketDao.saveOrUpdate(market);
        createAndSaveBetAction(market, new Date(), of("k1", "v1", "k2", "v2"), BET_ID);
        createAndSaveBetAction(market, new Date(), of("k1", "v1", "k2", "v2"), BET_ID + 1);
        assertThat(betActionDao.getBetActions(OptionalInt.of(1)).size(), is(1));
        assertThat(betActionDao.getBetActions(OptionalInt.of(2)).size(), is(2));
    }

    @Test
    public void testBetActionByBetId() {
        createMarketWithSingleAction();
        assertTrue(betActionDao.getBetAction(BET_ID).isPresent());
        assertFalse(betActionDao.getBetAction(BET_ID + 1).isPresent());
    }


    @Test
    public void testBetActionDateByBetId() {
        createMarketWithSingleAction();
        assertTrue(betActionDao.getBetActionDate(BET_ID).isPresent());
        assertFalse(betActionDao.getBetActionDate(BET_ID + 1).isPresent());
    }

    @Test
    public void testMarketPrices() {
        createMarketWithSingleAction();
        var betAction = betActionDao.getBetAction(BET_ID).get();
        assertThat(betAction.getMarketPrices(), notNullValue());
    }

    @Test(expected = LazyInitializationException.class)
    public void testMarketPricesLazy() {
        createMarketWithSingleAction();
        var betAction = betActionDao.getBetAction(BET_ID).get();
        betAction.getMarketPrices().getReciprocal(Side.BACK);
    }

    @Test
    public void testMarketPricesLazyFetch() {
        createMarketWithSingleAction();
        var action = betActionDao.getBetAction(BET_ID).get();
        betActionDao.fetchMarketPrices(action);
        assertThat(action.getMarketPrices().getTime(), notNullValue());
        assertThat(action.getMarketPrices().getReciprocal(Side.BACK).getAsDouble(), is(0.8333333333333333d));
        assertEquals(1.2d, action.getMarketPrices().getOverround(Side.BACK).getAsDouble(), 0.0001d);
    }

    @Test(expected = LazyInitializationException.class)
    public void testMarketLazy() {
        createMarketWithSingleAction();
        var action = betActionDao.getBetAction(BET_ID).get();
        action.getMarketPrices().getMarket().getName();
    }

    @Test
    public void testRunnerCount() {
        var market = newMarket("33", new Date(), CoreTestFactory.MATCH_ODDS);
        marketDao.saveOrUpdate(market);
        createAndSaveBetAction(market, addHours(new Date(), -1), PROPS, BET_ID);
        var stored = betActionDao.getBetAction(BET_ID).get();
        assertThat(stored.getMarket().getRunners().size(), is(market.getRunners().size()));
    }

    @Test
    public void testBetActionSortAsc() {
        saveActionsAndCheckOrder(comparing(BetAction::getActionDate));
    }

    @Test
    public void testBetActionSortDesc() {
        saveActionsAndCheckOrder(comparing(BetAction::getActionDate).reversed());
    }

    private void saveActionsAndCheckOrder(Comparator<BetAction> comparator) {
        var market = newMarket("33", new Date(), CoreTestFactory.MATCH_ODDS);
        marketDao.saveOrUpdate(market);
        var betActionEarlier = new BetAction(BetActionType.PLACE, DateUtils.addDays(new Date(), -1), new Price(2d, 30d, Side.LAY), market, CoreTestFactory.DRAW, BET_ID);
        var betActionLater = new BetAction(BetActionType.PLACE, new Date(), new Price(3d, 33d, Side.LAY), market, CoreTestFactory.DRAW, BET_ID + 1);
        var actions = List.of(betActionLater, betActionEarlier);
        Ordering.from(comparator).immutableSortedCopy(actions).forEach(betActionDao::saveOrUpdate);
        var betActionsForMarket = betActionDao.getBetActions(market.getId(), OptionalLong.empty(), empty());
        assertThat(betActionsForMarket.size(), is(2));
        assertThat(2d, is(betActionsForMarket.get(0).getPrice().getPrice()));
        assertThat(3d, is(betActionsForMarket.get(1).getPrice().getPrice()));
    }

    @Test
    public void testBetActionWithRunnerPrices() {
        var market = newMarket();
        var runnerPrices = RunnerPrices.create(232, List.of(new Price(2.3d, 22, Side.BACK)), 5d, 2.5d);
        var marketPrices = new MarketPrices(1, market, List.of(runnerPrices), new Date());
        marketPrices.setTime(DateUtils.addMonths(new Date(), -1));
        marketDao.saveOrUpdate(market);
        marketPricesDao.saveOrUpdate(marketPrices);
        var betAction = CoreTestFactory.newBetAction("1", market);
        betAction.setMarketPrices(marketPrices);
        betActionDao.saveOrUpdate(betAction);

        var actions = betActionDao.getBetActions(market.getId(), OptionalLong.empty(), empty());
        assertThat(actions.size(), is(1));
    }

    @Test
    public void testSharedPrices() {
        var market = newMarket();
        var runnerPrices = RunnerPrices.create(232, List.of(new Price(2.3d, 22, Side.BACK)), 5d, 2.5d);
        var marketPrices = new MarketPrices(1, market, List.of(runnerPrices), new Date());
        marketPrices.setTime(DateUtils.addMonths(new Date(), -1));
        marketDao.saveOrUpdate(market);
        marketPricesDao.saveOrUpdate(marketPrices);

        var betAction1 = CoreTestFactory.newBetAction("1", market);
        betAction1.setMarketPrices(marketPrices);
        betActionDao.saveOrUpdate(betAction1);

        var betAction2 = CoreTestFactory.newBetAction("2", market);
        betAction2.setMarketPrices(marketPrices);
        betActionDao.saveOrUpdate(betAction2);

        betActionDao.fetchMarketPrices(betAction1);
        betActionDao.fetchMarketPrices(betAction2);

        assertThat(betAction1.getMarketPrices().getId(), is(betAction2.getMarketPrices().getId()));
    }

}

File: mns-core/src/test/java/cz/fb/manaus/core/dao/BetActionDaoTest.kt
package cz.fb.manaus.core.dao

import com.google.common.collect.ImmutableMap.of
import com.google.common.collect.Ordering
import cz.fb.manaus.core.model.BetAction
import cz.fb.manaus.core.model.BetActionType
import cz.fb.manaus.core.model.Price
import cz.fb.manaus.core.model.Side
import cz.fb.manaus.core.test.CoreTestFactory
import cz.fb.manaus.core.test.CoreTestFactory.Companion.newMarket
import cz.fb.manaus.core.test.CoreTestFactory.Companion.newTestMarket
import cz.fb.manaus.core.test.ModelFactory
import org.apache.commons.lang3.time.DateUtils
import org.apache.commons.lang3.time.DateUtils.addHours
import org.hamcrest.CoreMatchers.hasItem
import org.hibernate.LazyInitializationException
import org.junit.Assert
import org.junit.Assert.assertThat
import org.junit.Test
import java.util.*
import java.util.Collections.singletonMap
import java.util.Optional.empty
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class BetActionDaoTest : AbstractDaoTest() {

    @Test
    fun `get action IDs for given market`() {
        val market = newTestMarket()
        marketDao.saveOrUpdate(market)
        createAndSaveBetAction(market, Date(), singletonMap("k1", "XXX"), AbstractDaoTest.BET_ID)
        val ids = betActionDao.getBetActionIds(market.id, OptionalLong.empty(), empty())
        assertThat(ids, hasItem(AbstractDaoTest.BET_ID))
    }

    @Test
    fun `update bet id`() {
        val market = newTestMarket()
        marketDao.saveOrUpdate(market)
        createAndSaveBetAction(market, Date(), emptyMap(), AbstractDaoTest.BET_ID)
        val newId = AbstractDaoTest.BET_ID + "_1"
        assertEquals(1, betActionDao.updateBetId(AbstractDaoTest.BET_ID, newId))
        assertEquals(0, betActionDao.updateBetId(AbstractDaoTest.BET_ID, newId))
        assertTrue(betActionDao.getBetAction(newId).isPresent)
        assertFalse(betActionDao.getBetAction(AbstractDaoTest.BET_ID).isPresent)
    }

    @Test
    fun `set bet id`() {
        val market = newTestMarket()
        marketDao.saveOrUpdate(market)
        val action = createAndSaveBetAction(market, Date(), emptyMap(), null)
        val actionId = action.id
        assertEquals(1, betActionDao.setBetId(actionId!!, "111"))
        assertEquals("111", betActionDao.get(actionId).get().betId)
    }

    @Test
    fun `get criteria`() {
        val market = newTestMarket()
        marketDao.saveOrUpdate(market)
        createAndSaveBetAction(market, Date(), singletonMap("k1", "XXX"), AbstractDaoTest.BET_ID)
        checkCount(market.id, OptionalLong.empty(), empty(), 1)
        checkCount(market.id, OptionalLong.of(CoreTestFactory.DRAW), empty(), 1)
        checkCount(market.id, OptionalLong.of(CoreTestFactory.DRAW + 1), empty(), 0)
        checkCount(market.id, OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY), 1)
        checkCount(market.id, OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.BACK), 0)
        checkCount(market.id, OptionalLong.empty(), Optional.of(Side.LAY), 1)
        checkCount(market.id, OptionalLong.empty(), Optional.of(Side.BACK), 0)
    }

    private fun checkCount(marketId: String, selId: OptionalLong, side: Optional<Side>, expectedCount: Int) {
        assertEquals(expectedCount, betActionDao.getBetActions(marketId, selId, side).size)
        assertEquals(expectedCount, betActionDao.getBetActionIds(marketId, selId, side).size)
    }

    @Test
    fun `action properties`() {
        val market = newMarket("33", Date(), CoreTestFactory.MATCH_ODDS)
        marketDao.saveOrUpdate(market)

        createAndSaveBetAction(market, Date(), singletonMap("k1", "newer"), AbstractDaoTest.BET_ID)
        val action = betActionDao.getBetActions("33", OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY))[0]
        betActionDao.getBetActions(OptionalInt.empty())
        assertEquals(1, action.properties.size)
        assertEquals("newer", action.properties["k1"])
    }


    @Test
    fun `action properties order`() {
        val market = newMarket("33", Date(), CoreTestFactory.MATCH_ODDS)
        marketDao.saveOrUpdate(market)
        createAndSaveBetAction(market, Date(), singletonMap("k1", "newer"), AbstractDaoTest.BET_ID)
        createAndSaveBetAction(market, addHours(Date(), -1), singletonMap("k1", "older"), AbstractDaoTest.BET_ID + 1)
        var action = betActionDao.getBetActions("33", OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY))[0]
        assertEquals(1, action.properties.size)
        assertEquals("older", action.properties["k1"])

        action = betActionDao.getBetActions("33", OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY))[1]
        assertEquals(1, action.properties.size)
        assertEquals("newer", action.properties["k1"])
        // delete with properties
        marketDao.delete(market.id)
    }

    @Test
    fun `action with properties delete`() {
        val market = newMarket("33", Date(), CoreTestFactory.MATCH_ODDS)
        marketDao.saveOrUpdate(market)
        createAndSaveBetAction(market, addHours(Date(), -1), singletonMap("k1", "older"), AbstractDaoTest.BET_ID)

        assertTrue(betActionDao.getBetAction(AbstractDaoTest.BET_ID).isPresent)
        marketDao.delete(market.id)
        assertFalse(betActionDao.getBetAction(AbstractDaoTest.BET_ID).isPresent)
    }

    @Test
    fun `no duplicate actions with multiple properties in result set`() {
        val market = newMarket("33", Date(), CoreTestFactory.MATCH_ODDS)
        marketDao.saveOrUpdate(market)
        createAndSaveBetAction(market, Date(), of("k1", "v1", "k2", "v2"), AbstractDaoTest.BET_ID)
        assertEquals(1, betActionDao.getBetActions(OptionalInt.empty()).size)
        assertEquals(1, betActionDao.getBetActions("33", OptionalLong.of(CoreTestFactory.DRAW), Optional.of(Side.LAY)).size)
    }

    @Test
    fun `maximal results limit`() {
        val market = newMarket("33", Date(), CoreTestFactory.MATCH_ODDS)
        marketDao.saveOrUpdate(market)
        createAndSaveBetAction(market, Date(), of("k1", "v1", "k2", "v2"), AbstractDaoTest.BET_ID)
        createAndSaveBetAction(market, Date(), of("k1", "v1", "k2", "v2"), AbstractDaoTest.BET_ID + 1)
        assertEquals(1, betActionDao.getBetActions(OptionalInt.of(1)).size)
        assertEquals(2, betActionDao.getBetActions(OptionalInt.of(2)).size)
    }

    @Test
    fun `get action by bet ID`() {
        createMarketWithSingleAction()
        assertTrue(betActionDao.getBetAction(AbstractDaoTest.BET_ID).isPresent)
        assertFalse(betActionDao.getBetAction(AbstractDaoTest.BET_ID + 1).isPresent)
    }


    @Test
    fun `get action date by bet ID`() {
        createMarketWithSingleAction()
        assertTrue(betActionDao.getBetActionDate(AbstractDaoTest.BET_ID).isPresent)
        assertFalse(betActionDao.getBetActionDate(AbstractDaoTest.BET_ID + 1).isPresent)
    }

    @Test
    fun `market prices field`() {
        createMarketWithSingleAction()
        val betAction = betActionDao.getBetAction(AbstractDaoTest.BET_ID).get()
        assertNotNull(betAction.marketPrices)
    }

    @Test(expected = LazyInitializationException::class)
    fun `lazy fetching market prices - negative`() {
        createMarketWithSingleAction()
        val betAction = betActionDao.getBetAction(AbstractDaoTest.BET_ID).get()
        betAction.marketPrices.getReciprocal(Side.BACK)
    }

    @Test
    fun `lazy fetching market prices - positive`() {
        createMarketWithSingleAction()
        val action = betActionDao.getBetAction(AbstractDaoTest.BET_ID).get()
        betActionDao.fetchMarketPrices(action)
        assertNotNull(action.marketPrices.time)
        assertEquals(0.8333333333333333, action.marketPrices.getReciprocal(Side.BACK).asDouble)
        Assert.assertEquals(1.2, action.marketPrices.getOverround(Side.BACK).asDouble, 0.0001)
    }

    @Test(expected = LazyInitializationException::class)
    fun `lazy fetching market - negative`() {
        createMarketWithSingleAction()
        val action = betActionDao.getBetAction(AbstractDaoTest.BET_ID).get()
        action.marketPrices.market.name
    }

    @Test
    fun `runner count`() {
        val market = newMarket("33", Date(), CoreTestFactory.MATCH_ODDS)
        marketDao.saveOrUpdate(market)
        createAndSaveBetAction(market, addHours(Date(), -1), AbstractDaoTest.PROPS, AbstractDaoTest.BET_ID)
        val stored = betActionDao.getBetAction(AbstractDaoTest.BET_ID).get()
        assertEquals(market.runners.size, stored.market.runners.size)
    }

    @Test
    fun `no impact of save order - asc`() {
        saveActionsAndCheckOrder(compareBy { it.actionDate })
    }

    @Test
    fun `no impact of save order - desc`() {
        saveActionsAndCheckOrder(compareByDescending { it.actionDate })
    }

    private fun saveActionsAndCheckOrder(comparator: Comparator<BetAction>) {
        val market = newMarket("33", Date(), CoreTestFactory.MATCH_ODDS)
        marketDao.saveOrUpdate(market)
        val earlier = ModelFactory.newAction(BetActionType.PLACE, DateUtils.addDays(Date(), -1), Price(2.0, 30.0, Side.LAY), market, CoreTestFactory.DRAW)
        earlier.betId = AbstractDaoTest.BET_ID
        val later = ModelFactory.newAction(BetActionType.PLACE, Date(), Price(3.0, 33.0, Side.LAY), market, CoreTestFactory.DRAW)
        later.betId = AbstractDaoTest.BET_ID + 1
        val actions = listOf(later, earlier)
        Ordering.from(comparator).immutableSortedCopy<BetAction>(actions).forEach { betActionDao.saveOrUpdate(it) }
        val betActionsForMarket = betActionDao.getBetActions(market.id, OptionalLong.empty(), empty())
        assertEquals(2, betActionsForMarket.size)
        assertEquals(2.0, betActionsForMarket[0].price.price)
        assertEquals(3.0, betActionsForMarket[1].price.price)
    }

    @Test
    fun `no duplicates due to multiple runner prices`() {
        val market = newTestMarket()
        val runnerPrices = ModelFactory.newRunnerPrices(232, listOf(Price(2.3, 22.0, Side.BACK)), 5.0, 2.5)
        val marketPrices = ModelFactory.newPrices(1, market, listOf(runnerPrices), Date())
        marketPrices.time = DateUtils.addMonths(Date(), -1)
        marketDao.saveOrUpdate(market)
        marketPricesDao.saveOrUpdate(marketPrices)
        val betAction = CoreTestFactory.newBetAction("1", market)
        betAction.marketPrices = marketPrices
        betActionDao.saveOrUpdate(betAction)

        val actions = betActionDao.getBetActions(market.id, OptionalLong.empty(), empty())
        assertEquals(1, actions.size)
    }

    @Test
    fun `market prices entity is shared by 2 actions`() {
        val market = newTestMarket()
        val runnerPrices = ModelFactory.newRunnerPrices(232, listOf(Price(2.3, 22.0, Side.BACK)), 5.0, 2.5)
        val marketPrices = ModelFactory.newPrices(1, market, listOf(runnerPrices), Date())
        marketPrices.time = DateUtils.addMonths(Date(), -1)
        marketDao.saveOrUpdate(market)
        marketPricesDao.saveOrUpdate(marketPrices)

        val betAction1 = CoreTestFactory.newBetAction("1", market)
        betAction1.marketPrices = marketPrices
        betActionDao.saveOrUpdate(betAction1)

        val betAction2 = CoreTestFactory.newBetAction("2", market)
        betAction2.marketPrices = marketPrices
        betActionDao.saveOrUpdate(betAction2)

        betActionDao.fetchMarketPrices(betAction1)
        betActionDao.fetchMarketPrices(betAction2)

        assertEquals(betAction1.marketPrices.id, betAction2.marketPrices.id)
    }
}